### PR TITLE
Test branch

### PR DESCRIPTION
**Pull request by anu3990**
So we can incorporate your pull request, please share the following:
* What issue are you addressing with this pull request?
https://project.marklogic.com/jira/browse/JAVA-19,
https://project.marklogic.com/jira/browse/JAVA-20,
https://project.marklogic.com/jira/browse/JAVA-21 and 
https://project.marklogic.com/jira/browse/JAVA-22
* Are you modifying the correct branch? (See CONTRIBUTING.md)
* Have you run unit tests? (See CONTRIBUTING.md)
* Version of MarkLogic Java Client API (see Readme.txt)
* Version of MarkLogic Server (see admin gui on port 8001)
* Java version (`java -version`)
* OS and version
* What Changed: What happened before this change? What happens without this change?

### CODE REVIEW

File: marklogic-client-api/src/main/java/com/marklogic/client/impl/HTTPSamlAuthInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	    }
+	 
+	    String samlHeaderValue = RESTServices.AUTHORIZATION_TYPE_SAML+ " "+ RESTServices.AUTHORIZATION_PARAM_TOKEN + "=" + authorizationTokenValue;
+	    Request authenticatedRequest = request.newBuilder().header(RESTServices.HEADER_AUTHORIZATION, samlHeaderValue).build();
+	    return chain.proceed(authenticatedRequest);
+	  }
+	  
+	  private synchronized void authorize(Instant expiry) {
+		  if(expiringSAMLAuth == null)
+			  throw new MarkLogicInternalException("ExpiringSAMLAuth cannot be null");
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/impl/HTTPSamlAuthInterceptor.java]: On the first call, both the expiry param and the expiringSAMLAuth field will be null.  

So, I believe this test will cause the first call to fail.

Instead, if the expiry param is null but the expiringSAMLAuth field is not null, some other thread must have updated the expiringSAMLAuth field, so the authorize() method should return.

File: marklogic-client-api/src/main/java/com/marklogic/client/impl/HTTPSamlAuthInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		  expiringSAMLAuth = authorizer.apply(expiringSAMLAuth);
+		  if(expiringSAMLAuth==null)
+			  throw new IllegalArgumentException("SAML Authentication cannot be null");
+		  
+		  if(expiringSAMLAuth.getAuthorizationToken() == null)
+			  throw new IllegalArgumentException("SAML Authentication token cannot be null");
+		  authorizationTokenValue = expiringSAMLAuth.getAuthorizationToken();
+		  
+		  if(expiringSAMLAuth.getExpiry() == null)
+			  throw new IllegalArgumentException("SAML authentication does not have expiry value.");
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/impl/HTTPSamlAuthInterceptor.java]: One other validation occurs to partially addressing the issue that you raised in our meeting yesterday -- if the expiry instant .isBefore(Instant.now), the token has already expired.

File: marklogic-client-api/src/main/java/com/marklogic/client/impl/OkHttpServices.java [hunk 3]
```diff
@@ -238,45 +237,138 @@ private FailedRequest extractErrorFields(Response response) {
...
-
+  public void connect(String host, int port, String database, SecurityContext securityContext){
+	String user = null;
+	Map<String, String> kerberosOptions = null;
+	String password = null;
+	Authentication type = null;
+	SSLContext sslContext = null;
+	SSLHostnameVerifier sslVerifier = null;
+	X509TrustManager trustManager = null;
+	String authorizationTokenValue = null;
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/impl/OkHttpServices.java]: Can we remove some of these variables if we move the contents of the branches into methods?

Would the code be easier to read and thus understand if the indentation of the variables that remain aligned with line 249?

File: marklogic-client-api/src/main/java/com/marklogic/client/impl/OkHttpServices.java [hunk 4]
```diff
@@ -238,45 +237,138 @@ private FailedRequest extractErrorFields(Response response) {
...
+			if (basicContext.getTrustManager() != null)
+				trustManager = basicContext.getTrustManager();
+			if (basicContext.getSSLHostnameVerifier() != null) {
+				sslVerifier = basicContext.getSSLHostnameVerifier();
+			} else {
+				sslVerifier = SSLHostnameVerifier.COMMON;
+			}
+			interceptor = new HTTPBasicAuthInterceptor(credentials);
+	        checkFirstRequest = false;
+		}
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/impl/OkHttpServices.java]: If I recall correctly, an earlier version moved the SSL setup out of the branches and moved the other code from the branches into methods called from the branch by casting the security context and passing the client builder. 

Would that approach still work?

Comment (anu3990) [reply marklogic-client-api/src/main/java/com/marklogic/client/impl/OkHttpServices.java]: Included configureAuthentication overloads. Sorry it was missed during the branch change.

File: marklogic-client-api/src/main/java/com/marklogic/client/DatabaseClientFactory.java [hunk 5]
```diff
@@ -509,6 +520,88 @@ public KerberosAuthContext withSSLHostnameVerifier(SSLHostnameVerifier verifier)
...
+					public Instant getExpiry() {
+						return expiry;
+					}
+					
+					@Override
+					public String getAuthorizationToken() {
+						return authorizationToken;
+					}
+				};
+	        }
```
[the remark this answers is not in the document]
Comment (anu3990) [reply marklogic-client-api/src/main/java/com/marklogic/client/DatabaseClientFactory.java]: Yes that is correct. Including the change.

File: marklogic-client-api/src/main/java/com/marklogic/client/impl/OkHttpServices.java (hunk 4 above)
Comment (ehennum) [reply marklogic-client-api/src/main/java/com/marklogic/client/impl/OkHttpServices.java]: Would it work to pass the SecurityContext and OkHttpClient.Builder to each configureAuthentication() method?

As in

    if (securityContext instanceof BasicAuthContext) {
        clientBldr = configureAuthentication((BasicAuthContext) securityContext, clientBldr);
    } else  (securityContext instanceof DigestAuthContext) {
        clientBldr = configureAuthentication((DigestAuthContext) securityContext, clientBldr);
    ... other tests and casts on calls ...

In other words, to move the code that accesses the security context into the configureAuthentication() methods?

Also, given that the SSL getters are defined on SecurityContext, could the SSL configuration be performed once outside of the branches on the specific instances of SecurityContext?

Is that possible?

**Comment by anu3990** [comment]
Added copyright and javadoc corresponding to https://project.marklogic.com/jira/browse/JAVA-27 and https://project.marklogic.com/jira/browse/JAVA-28
